django/db/migrations/executor.py

```diff
@@ -200,10 +200,10 @@ def _migrate_all_backwards(self, plan, full_plan, fake):
                 # process.
                 break
             if migration in migrations_to_run:
+                # The state before this migration
+                states[migration] = state.clone()
                 if "apps" not in state.__dict__:
                     state.apps  # Render all -- performance critical
-                # The state before this migration
-                states[migration] = state
                 # The old state keeps as-is, we continue with the new state
                 state = migration.mutate_state(state, preserve=True)
```

Comment:
We may no longer need to set `preserve=True` here since state is cloned already.

Reply:
I'm not sure this is adequate, but my migration understanding is rusty.

For migrations that are planed to be unapplied (`migration in migrations_to_run`) we do want their rendered state to be used below when calling `self.unapply_migration` as it will inevitably result in `database_forwards` being called for each operation and the latter still relies on `.state` to perform schema changes (see ticket-29898).

Is the speedup you're experiencing due to the fact that the `state_forwards` calls performed by `Migration.unapply` for each of its operations are now cheaper because the apps are not invalidated?

From what I understand the changes you are proposing does
1. Avoid eager model rendering for the last migration that needs to be unapplied
2. Make `state_forwards` and state cloning within `Migration.unapply` for each operation really cheap for this particular migration
3. Cause `O=len(last_migration.operations)` model rendering from scratch for each `database_backwards` call

I think this change might make unapplying a single migration with a single operation faster at the expense of slowing down other scenarios. Does that add up with your testing?

Reply:
@charettes Ah that explains why we render apps on state 👍

The slowdown was _only_ caused by the block that mutates solely for post-migration state:

```python
        # Generate the post migration state by starting from the state before
        # the last migration is unapplied and mutating it to include all the
        # remaining applied migrations.
        last_unapplied_migration = plan[-1][0]
        state = states[last_unapplied_migration]
        del state.apps
        for index, (migration, _) in enumerate(full_plan):
            if migration == last_unapplied_migration:
                for migration, _ in full_plan[index:]:
                    if migration in applied_migrations:
                        migration.mutate_state(state, preserve=False)
                break
```


As you pointed out, we should indeed leave it rendered for this:

```python
        for migration, _ in plan:
            self.unapply_migration(states[migration], migration, fake=fake)
            applied_migrations.remove(migration)
```

(I confirmed this by when testing reversing to `zero`: my patch showed a significant slow down as it's my largest migration.) 


So rethinking this, the changes should be isolated to that post-migration mutating block.  I think this is more appropriate:

```diff
--- a/django/db/migrations/executor.py
+++ b/django/db/migrations/executor.py
@@ -224,6 +224,7 @@ class MigrationExecutor:
         # remaining applied migrations.
         last_unapplied_migration = plan[-1][0]
         state = states[last_unapplied_migration]
+        del state.apps
         for index, (migration, _) in enumerate(full_plan):
             if migration == last_unapplied_migration:
                 for migration, _ in full_plan[index:]: 
```

^ testing that change fixes the issue of reversing the last migration and doesn't affect reversing to `zero`.